{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "41a5c995-a3fc-41fb-a646-f73a716f44dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam\n",
    "import glob\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb7ec503-fd7d-4dc8-9c4b-e0d2df2f350a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "238e503e-bb6a-4c7c-882d-bc1671b6ee76",
   "metadata": {},
   "outputs": [],
   "source": [
    "title = \"Histgram of 1083 Insert Region Reads \" + key_ID\n",
    "def Histplot_Save(_plot_data, _title, _path):\n",
    "    # Example code to create and save the histogram\n",
    "    plt.figure(figsize=(10, 6))\n",
    "    ax = sns.histplot(data = _plot_data, x = \"Length\", hue='Type', element='step', \n",
    "         log_scale=(False, True), bins=region_end - region_start)\n",
    "    # Add a title\n",
    "    ax.set_title(_title)\n",
    "    output_file = _path.replace(\".bam\", \"_hist.png\")\n",
    "    plt.savefig(output_file, dpi=100, bbox_inches='tight')\n",
    "    plt.close()\n",
    "    return None\n",
    "\n",
    "def Generate_INDEL_STAT_For_Alignment_In_Given_Region(_bam_path, _chr, _start, _end):\n",
    "    \"\"\"\n",
    "    only using sequences inside certain region, and return Insert, delete & Match for each alignment\n",
    "    \"\"\"\n",
    "    # Iterate through each read in the BAM file\n",
    "    bam_file = pysam.AlignmentFile(_bam_path, \"rb\")\n",
    "    chromosome = _chr\n",
    "    region_start = _start\n",
    "    region_end = _end\n",
    "\n",
    "    # Initialize counters\n",
    "    match_lengths = []\n",
    "    insertion_lengths = []\n",
    "    deletion_lengths = []\n",
    "    sub_lengths = []\n",
    "    query_names = []\n",
    "    \n",
    "    for read in bam_file.fetch():\n",
    "        match_length = 0\n",
    "        insertion_length = 0\n",
    "        deletion_length = 0\n",
    "        sub_length = 0\n",
    "        # Skip unmapped reads\n",
    "        if read.is_unmapped:\n",
    "            continue\n",
    "    \n",
    "        query_names.append(read.qname)\n",
    "        # Initialize total lengths for the current read\n",
    "        total_match_length = 0\n",
    "        total_insertion_length = 0\n",
    "        total_deletion_length = 0\n",
    "    \n",
    "        # Current genomic position as the read is aligned\n",
    "        read_genomic_pos = read.reference_start\n",
    "    \n",
    "        # Process CIGAR tuples to aggregate INDELs and match lengths for the read\n",
    "        for cigar_tuple in read.cigartuples:\n",
    "            operation = cigar_tuple[0]\n",
    "            length = cigar_tuple[1]\n",
    "    \n",
    "            # Calculate the next genomic position based on the current CIGAR operation\n",
    "            if operation in [0, 7, 8]:  # Match, sequence match, sequence mismatch (M, =, X)\n",
    "                region_overlap_start = max(read_genomic_pos, region_start)\n",
    "                region_overlap_end = min(read_genomic_pos + length, region_end)\n",
    "    \n",
    "                # If the CIGAR operation overlaps the region of interest\n",
    "                if region_overlap_start < region_overlap_end:\n",
    "                    # Add to the total match length for this read\n",
    "                    total_match_length += region_overlap_end - region_overlap_start\n",
    "                read_genomic_pos += length\n",
    "    \n",
    "            elif operation == 1:  # Insertion\n",
    "                # Count the insertion if it occurs within the region\n",
    "                if region_start <= read_genomic_pos < region_end:\n",
    "                    total_insertion_length += length\n",
    "    \n",
    "            elif operation == 2:  # Deletion\n",
    "                region_overlap_start = max(read_genomic_pos, region_start)\n",
    "                region_overlap_end = min(read_genomic_pos + length, region_end)\n",
    "    \n",
    "                # If the deletion overlaps the region\n",
    "                if region_overlap_start < region_overlap_end:\n",
    "                    total_deletion_length += region_overlap_end - region_overlap_start\n",
    "                read_genomic_pos += length\n",
    "    \n",
    "            elif operation in [3, 4, 5]:  # Skips, soft clips, hard clips (N, S, H)\n",
    "                if operation == 3:  # Skip (N) also moves the genomic position\n",
    "                    read_genomic_pos += length\n",
    "    \n",
    "        # Store the aggregated lengths for the current read\n",
    "        if total_match_length >= 0:\n",
    "            match_lengths.append(total_match_length)\n",
    "        if total_insertion_length >= 0:\n",
    "            insertion_lengths.append(total_insertion_length)\n",
    "        if total_deletion_length >= 0:\n",
    "            deletion_lengths.append(total_deletion_length)# Close the BAM file\n",
    "    bam_file.close()\n",
    "    \n",
    "    match_data = {\n",
    "        'Type': ['Match'] * len(match_lengths) + ['Ins'] * len(insertion_lengths) + \n",
    "        ['Del'] * len(deletion_lengths), # + ['Sub'] * len(sub_lengths),\n",
    "        'Length': match_lengths + insertion_lengths + deletion_lengths# + sub_lengths\n",
    "    }\n",
    "    \n",
    "    df_CIGAR_sum = pd.DataFrame(data = {\"qname\": query_names, \"Match\":match_lengths, \"Ins\":insertion_lengths,\n",
    "                        \"Del\": deletion_lengths})\n",
    "    return match_data, df_CIGAR_sum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2748d743-b6fe-45fc-b234-f0f30a62b7bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['1083_Overlap_AAVS1_edited_1.bam',\n",
       " '1083_Overlap_AAVS1_edited_2.bam',\n",
       " '1083_Overlap_AAVS1_edited_3.bam',\n",
       " '1083_Overlap_AAVS1_edited_4.bam']"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bam_files = sorted(glob.glob(\"*.bam\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9819a374-342a-4bbc-a89e-813f9387d631",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "19c0c362-2600-4789-a7cb-bf10bed930a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_bam_path = \"1083_Overlap_Reads_AAVS1_edited_4.bam\"\n",
    "\n",
    "_path_excel = \"1083_insert_INDEL_Analysis.xlsx\"\n",
    "with pd.ExcelWriter(_path_excel) as writer:\n",
    "    for input_bam_path in bam_files:\n",
    "        key_ID = input_bam_path.replace(\".bam\", \"\").replace(\"1083_Overlap_\", \"\")\n",
    "        if (\"AAVS1\" in key_ID):\n",
    "            chromosome, region_start, region_end = \"AAVS1_edited_1\", 221, 1304\n",
    "        elif(\"PRNP\" in key_ID):\n",
    "            chromosome, region_start, region_end = \"PRNP_edited\", 185, 1268\n",
    "        elif(\"HEK3\" in key_ID):\n",
    "            chromosome, region_start, region_end = \"HEK3_edited\", 136, 1219\n",
    "        elif(\"IDS\" in key_ID):\n",
    "            chromosome, region_start, region_end = \"IDS_edited\", 258, 1341\n",
    "        \n",
    "        plot_data, df_STAT = Generate_INDEL_STAT_For_Alignment_In_Given_Region(input_bam_path, chromosome, region_start, region_end)\n",
    "        df_STAT.to_excel(writer, sheet_name = key_ID, index=None,\n",
    "                         freeze_panes=(1, 1))\n",
    "        Histplot_Save(plot_data, \"Histgram of 1083 Insert Region Reads \" + key_ID, input_bam_path)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
